brainiak/eventseg/event.py

```diff
@@ -489,3 +538,143 @@ def model_prior(self, t):
+        iteration_var : float
+            Current variance in simulated annealing
+
+        curr_ll: float
+            Log-likelihood of current model
+
+        Returns
+        -------
+        return_ll : ndarray with length equal to length of X
+            Log-likelihood after merge/split (same as curr_ll if no
+            merge/split improved curr_ll)
+
+        return_lg : list of time by event ndarrays
+            Log probability of each timepoint belonging to each event,
+            for each dataset (same as log_gamma if no merge/split
+            improved curr_ll)
+
+        return_mp : voxel by event ndarray
+            Mean patterns of events (after possible merge/split)
+        """
+
+        # Compute current probabilities and mean patterns
+        n_train = len(X)
+        n_dim = X[0].shape[0]
+
+        seg_prob = [np.exp(lg) / np.sum(np.exp(lg), axis=0)
+                    for lg in log_gamma]
+        mean_pat = np.empty((n_train, n_dim, self.n_events))
+        for i in range(n_train):
+            mean_pat[i, :, :] = X[i].dot(seg_prob[i])
+        mean_pat = np.mean(mean_pat, axis=0)
+
+        # Compute correlations between every pair of neighboring
+        # events and their merged pattern. High values of
+        # merge_corr indicate events that are good candidates for
+        # being merged.
+        # Also, compute correlations between every event and each half of
+        # the event, with the midpoint defined as the point where
+        # the cumulative probability (across time) of being in the
+        # event = 0.5. Low values of split_corr indicate events
+        # that are good candidates for being split.
+        merge_corr = np.zeros(self.n_events)
+        merge_pat = np.empty((n_train, n_dim, self.n_events))
+        split_corr = np.zeros(self.n_events)
+        split_pat = np.empty((n_train, n_dim, 2 * self.n_events))
+        for i, sp in enumerate(seg_prob):  # Iterate over datasets
+            merge = np.zeros((sp.shape[0], sp.shape[1]))
+            split = np.zeros((sp.shape[0], 2 * sp.shape[1]))
+            cs = np.cumsum(sp, axis=0)
+            for e in range(sp.shape[1]):
+                mid = np.where(cs[:, e] >= 0.5)[0][0]
+                cs_first = cs[mid, e] - sp[mid, e]
+                cs_second = 1 - cs_first
+                split[:mid, 2 * e] = sp[:mid, e] / cs_first
+                split[mid:, 2 * e + 1] = sp[mid:, e] / cs_second
+                merge[:, e] = sp[:, e:(e + 2)].mean(1)
+            merge_pat[i, :, :] = X[i].dot(merge)
+            split_pat[i, :, :] = X[i].dot(split)
+        merge_pat = np.mean(merge_pat, axis=0)
+        split_pat = np.mean(split_pat, axis=0)
+        for e in range(self.n_events):
+            split_corr[e] = np.corrcoef(mean_pat[:, e],
+                                        split_pat[:, (2 * e):(2 * e + 2)],
+                                        rowvar=False)[0, 1:3].max()
+            merge_corr[e] = np.corrcoef(merge_pat[:, e],
```

Comment:
I'm curious -- why take the maximum split correlation and the minimum merge correlation? Why not average?

Reply:
The goal is to select the merge and split combination that is most likely to increase the log likelihood, so we want to find the pair of events that is most similar to merge, and the split halves of an event that are most dissimilar to split - I've added more comments now.